### PR TITLE
[LLM] Optimize kv_cache for gptj model family

### PR DESCRIPTION
**Pull request by sgwhat**
## Description
In this PR, we refactored the code for `key` and `value` inside the GPTJAttention forward function to optimize the model generation of reset tokens' latency.

### 1. Why the change?
The rest tokens latency need to be optimized.

### 2. User API changes
N/A

### 3. Summary of the change 
- add `create_kv_cache` and `append_kv_cache` to instead of `torch.cat`.
- refactor new functionality of `gptj_attention_forward`.

### 4. How to test?
- [x] Unit test

### CODE REVIEW

File: python/llm/src/bigdl/llm/transformers/convert.py [hunk 1]
```diff
@@ -173,6 +173,15 @@ def optimize(model):
                         module.SelfAttention,
                         chatglm_attention_forward
                         )
+    elif "gptj" in model.config.model_type:
+        modeling_module_name = model.__class__.__module__
+        module = importlib.import_module(modeling_module_name)
+        if "GPTJForCausalLM" in model.config.architectures:
```
Comment (jason-dai) [inline python/llm/src/bigdl/llm/transformers/convert.py:179]: do we need this? it seems true for all gpt-j models (e.g., https://huggingface.co/EleutherAI/gpt-j-6b)

Comment (sgwhat) [reply python/llm/src/bigdl/llm/transformers/convert.py]: removed.

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+    return (tensor * cos) + (rotate_every_two(tensor) * sin)
+
+
+def rotate_every_two(x: torch.Tensor) -> torch.Tensor:
+    x1 = x[:, :, :, ::2]
+    x2 = x[:, :, :, 1::2]
+    x = torch.stack((-x2, x1), dim=-1)
+    return x.flatten(-2)  # in einsum notation: rearrange(x, '... d j -> ... (d j)')
+
+
```
[the remark this answers is not in the document]
Comment (sgwhat) [reply python/llm/src/bigdl/llm/transformers/models/gptj.py:43]: I think the implementation of gptj is difference with llama and baichuan, but we could make this done with a new functionality.

[the remark this answers is not in the document]
Comment (jason-dai) [reply python/llm/src/bigdl/llm/transformers/models/gptj.py:43]: what's exactly why we have a `model_family` parameter in `bigdl.llm.transformers.models.utils.apply_rotary_pos_emb`

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+import torch
+from typing import Optional, Tuple, Union
+from bigdl.llm.transformers.models.utils import create_kv_cache, append_kv_cache, \
+    apply_rotary_pos_emb
+from transformers.utils.import_utils import is_torch_fx_proxy
+
+
+KV_CACHE_ALLOC_BLOCK_LENGTH = 256
+KV_CACHE_ALLOC_MIN_LENGTH = 512
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/gptj.py:28]: How is this value set?

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        q_pass = query[:, :, :, self.rotary_dim:]
+
+        q_rot, k_rot = apply_rotary_pos_emb(q_rot, k_rot, cos, sin, position_ids, "gptj")
+
+        key = torch.cat([k_rot, k_pass], dim=-1)
+        query = torch.cat([q_rot, q_pass], dim=-1)
+    else:
+        query, key = apply_rotary_pos_emb(query, k_rot, cos, sin, position_ids, "gptj")
+
+    batch_size, cur_length = query.shape[0], query.shape[1]
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/gptj.py:129]: shouldn't it be the length of key?

File: python/llm/src/bigdl/llm/transformers/models/gptj.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+                                                       dtype=query.dtype,
+                                                       device=device)
+            new_cache_k[:] = cache_k
+            new_cache_v[:] = cache_v
+            cache_k = new_cache_k
+            cache_v = new_cache_v
+        key, value = append_kv_cache(cache_k, cache_v, key, value)
+
+    elif use_cache:
+        max_cache_length = max(KV_CACHE_ALLOC_MIN_LENGTH, cur_length) \
```
Comment (hkvision) [inline python/llm/src/bigdl/llm/transformers/models/gptj.py:162]: is this correct?

File: python/llm/src/bigdl/llm/transformers/models/gptj.py (hunk 4 above)
Comment (hkvision) [reply python/llm/src/bigdl/llm/transformers/models/gptj.py:128]: use hidden states to get batch size

**Review by hkvision — APPROVED** [review]

**Pull request merged**